target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,461 @@
+	unsigned int cmu_en_ckoobs = 0x0;
+	unsigned int test_en = 0x0;
+
+	if (rate == 500000000 && parent_rate == 25000000 && pll->id == CLK_CPU) {
+		cmu_ncode_in = 0x74;
+		cmu_divn3_sel = 0x1;
+		cmu_big_kvco = 0x1;
+	} else if (rate == 300000000 && parent_rate == 25000000 && pll->id == CLK_CPU) {
+		cmu_ncode_in = 0x5C;
+		cmu_divn3_sel = 0x2;
+		cmu_big_kvco = 0x0;
+	} else {
+		pr_err("not allowed set clock %d to %luMHz\n", pll->id, rate / 1000000);
+		return 0;
+	}
+
+	ctrl0 = (cmu_fcode_in<<20)|(cmu_divn2<<12)|(cmu_ncode_in<<4)|
+		(cmu_bypass_pi<<3)|(cmu_sel_div4<<2)|(cmu_sel_prediv<<0);
+	ctrl1 = (cmu_divn3_sel<<27)|(cmu_divn2_selb<<26)|(cmu_en_ssc<<25)|
+		(cmu_step_in<<12)|(cmu_tbase_in<<0);
+	misc = (cmu_ssc_order<<31)|(cmu_time2_rst_width<<29)|(cmu_time0_ck<<26)|
+		(cmu_clkrdy<<24)|(cmu_big_kvco<<23)|(cmu_lpf_rs<<20)|
+		(cmu_en_center_in<<19)|(cmu_en_wd<<18)|(cmu_pi_i_sel<<14)|
+		(cmu_sel_cp_i<<10)|(cmu_sel_cco<<9)|(cmu_ldo_sel<<6)|
+		(cmu_lpf_cp<<5)|(cmu_cp_new_en<<4)|(cmu_ldo_en<<3)|
+		(cmu_vc_dly<<2)|(cmu_en_ckoobs<<1)|(test_en<<0);
+
+	spin_lock_irqsave(&pll->lock, flags);
+
+	sw_w32(sw_r32(RTL838X_PLL_GLB_CTRL) & ~(1 << 12), RTL838X_PLL_GLB_CTRL);
+	sw_w32(sw_r32(RTL838X_PLL_GLB_CTRL) & ~1, RTL838X_PLL_GLB_CTRL);
+
+	sw_w32(misc, RTL838X_PLL_CPU_MISC_CTRL);
+	sw_w32(ctrl0, RTL838X_PLL_CPU_CTRL0);
+	sw_w32(ctrl1, RTL838X_PLL_CPU_CTRL1);
+	
+	sw_w32(sw_r32(RTL838X_PLL_GLB_CTRL) | 1, RTL838X_PLL_GLB_CTRL);
+	while(!(sw_r32(RTL838X_PLL_GLB_CTRL) & (1 << 8)));
+	sw_w32(sw_r32(RTL838X_PLL_GLB_CTRL) | (1 << 12), RTL838X_PLL_GLB_CTRL);
+
+	spin_unlock_irqrestore(&pll->lock, flags);
+
+	return 0;
+}
+
+static long rtl838x_pll_round_rate(struct clk_hw *hw, unsigned long rate, unsigned long *parent_rate)
+{
+	struct rtl_pll *pll = rtl_hw_to_pll(hw);
+	if (pll->id == CLK_CPU) 
+		return (rate < 450000000) ? 300000000 : 500000000;
+	else if (pll->id == CLK_MEM) 
+		return (rate < 246875000) ? 193750000 : 300000000;
+	else
+		return 200000000;
+}
+
+static int rtl838x_pll_enable(struct clk_hw *hw)
+{
+	return 0;
+}
+
+static void rtl838x_pll_disable(struct clk_hw *hw)
+{
+}
+
+static int rtl838x_pll_is_enabled(struct clk_hw *hw)
+{
+	return 1;
+}
+
+static int rtl839x_pll_set_rate(struct clk_hw *hw, unsigned long rate, unsigned long parent_rate)
+{
+	return 0;
+}
+
+static long rtl839x_pll_round_rate(struct clk_hw *hw, unsigned long rate, unsigned long *parent_rate)
+{
+	return 0;
+}
+
+static int rtl839x_pll_enable(struct clk_hw *hw)
+{
+	return 0;
+}
+
+static void rtl839x_pll_disable(struct clk_hw *hw)
+{
+}
+
+static int rtl839x_pll_is_enabled(struct clk_hw *hw)
+{
+	return 1;
+}
+
+/*
+ * Initialization functions to register the clocks.
+ */
+
+static const struct clk_ops rtl838x_pll_ops = {
+	.round_rate = rtl838x_pll_round_rate,
+	.recalc_rate = rtl83xx_pll_recalc_rate,
+	.set_rate = rtl838x_pll_set_rate,
+	.enable = rtl838x_pll_enable,
+	.disable = rtl838x_pll_disable,
+	.is_enabled = rtl838x_pll_is_enabled
+};
```

Comment:
if you leave these undefined/unset; doesn't the clkframe work have some defaults/skip them? That also makes your above stubs not needed I think ...